storagenode/pieces/cache.go

```diff
@@ -0,0 +1,163 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package pieces
+
+import (
+	"context"
+	"sync"
+
+	"go.uber.org/zap"
+	"storj.io/storj/pkg/storj"
+	"storj.io/storj/storage"
+)
+
+// StoreWithCache has access to the pieces store and also keeps track of real time
+// totals of space used for all pieces and for all pieces by satellite
+type StoreWithCache struct {
```

Comment:
Do you think it's necessary to have this be a new type, as opposed to adding this functionality to the main piece store? It seems like most services shouldn't need to care whether their piece stores have space-used caches inside or not.

Reply:
Maybe we could have some kind of SpaceUsedGetter interface, and `pieces.Store` has one of those, and it abstracts how the space-used calculation happens, whether it's cached, how often the cache is updated, etc?

Reply:
originally i had it added to the main piece store, but I changed it this way based on Egon's comment here https://github.com/storj/storj/pull/2707#discussion_r310641344

im open to either way.

Reply:
Oh. Did he clarify to you what he meant outside of the PR? If he wants a wrapper around all of pieces.Store, maybe he has a rationale that hasn't occurred to me.

Reply:
I.e., I wonder if maybe he meant a wrapper around `SpaceUsed()` and `SpaceUsedInNamespace()` themselves?

Reply:
yeah i asked him what he meant by "wrapper" and he linked an example that isaac did that used embedding. So I believe he was saying to embed the Store in a wrapper that has caching.  Her is the example: https://github.com/storj/storj/pull/2681/files#diff-2196cdf1c7f5a0b80f2b17bf167406b5R23

Reply:
Huh. That seems wrong to me in this case. Maybe a wrapper around `filestore.Store` would make more sense? Or perhaps we should wait for @egonelbre to clarify once he's back online.

Reply:
yeah maybe i misunderstood something.  would be nice to get egon's take on this.

Reply:
Usually you want to wrap the smallest thing possible to add the functionality. This ensures that you don't have to forward calls that much.

The reason I suggested trying wrapping is that it might preserve the rest of the implementation clear of this concern. I.e. you would be able to remove / add / replace the caching with some other approach when needed. Similarly you can change the store without worrying breaking caching. In principle it keeps clarity between "this is what we cache", "this is how we store" and "this is how we do X / Z".